src/main/scala/com/fulcrumgenomics/fastq/DemuxFastqs.scala

```diff
@@ -0,0 +1,663 @@
+) extends FgBioTool with LazyLogging {
+
+  import DemuxFastqs._
+
+  // Check all input and output files are readable or writable respectively.
+  Io.assertReadable(fastq1 ++ fastq2 ++ fastq7 ++ fastq5)
+  Io.assertReadable(this.sampleSheet)
+  Io.assertWritableDirectory(this.output)
+  Io.assertCanWriteFile(this.metrics)
+
+  override def execute(): Unit = {
+    // Make some data structures to help us on our way
+    val r1ReadStructure          = ReadStructure(rs1)
+    val r2ReadStructure          = rs2.map(ReadStructure(_))
+    val i7ReadStructure          = rs7.map(ReadStructure(_))
+    val i5ReadStructure          = rs5.map(ReadStructure(_))
+
+    // Verify that the template bases are only on read one or read two
+    if (r1ReadStructure.template.isEmpty)           throw new IllegalArgumentException("No template bases found in the read structure for read one.")
+    if (r2ReadStructure.exists(_.template.isEmpty)) throw new IllegalArgumentException("No template bases found in the read structure for read two.")
+    if (i7ReadStructure.exists(_.template.nonEmpty)) throw new IllegalArgumentException("Template bases not allowed in the i7 read.")
+    if (i5ReadStructure.exists(_.template.nonEmpty)) throw new IllegalArgumentException("Template bases not allowed in the i5 read.")
+
+    val fastqs = Seq(fastq1, fastq2, fastq7, fastq5)
+    val readStructures: Seq[ReadStructure] = Seq(Some(r1ReadStructure), r2ReadStructure, i7ReadStructure, i5ReadStructure).flatten
+    if (readStructures.length != fastqs.count(_.nonEmpty)) throw new IllegalArgumentException("The # of read structures did not match the number of fastqs.")
+
+    // check the molecular barcodes and associated tags, if any exist.
+    val molecularBarcodes = readStructures.flatMap { rs => rs.molecularBarcode }
+    if (molecularBarcodes.isEmpty && multiUmiTags.nonEmpty) {
+      throw new IllegalArgumentException("No molecular barcodes found but molecular barcode SAM tags were given.")
+    }
+    if (multiUmiTags.length > 1 && multiUmiTags.length != molecularBarcodes.length) {
+      throw new IllegalArgumentException("Either one SAM tag or the same number as the molecular barcode segments should be given " +
+        s"(found ${molecularBarcodes.length} molecular barcodes but given ${multiUmiTags.length} tags."
+      )
+    }
+
+    // Read the sample sheet, convert to a custom sample type, and set the barcode bases for each sample
+    val sampleSheet: SampleSheet = DemuxFastqs.sampleSheet(sampleSheet=this.sampleSheet, lane=lane).sampleBarcodes(readStructures:_*)
+
+    // If the sample barcodes are in-line in the read (i.e. in r1 or r2) then they may be reverse complemented (ex. amplification).
+    if (matchReverseComplement) {
+      // check that we only have in-line sample barcodes.
+      if (Seq(i7ReadStructure, i5ReadStructure).exists(_.exists(_.sampleBarcode.nonEmpty))) {
```

Comment:
Again, this seems totally arbitrary - that you can't even _have_ i5 or i7 indices if you have inline indices which might be RC'd.